src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -370,126 +418,410 @@ private sealed class SendFileOperation : WriteOperation
                 SocketPal.TryCompleteSendFile(context._socket, FileHandle, ref Offset, ref Count, ref BytesTransferred, out ErrorCode);
         }
 
-        private enum QueueState
+        // In debug builds, this struct guards against:
+        // (1) Unexpected lock reentrancy, which should never happen
+        // (2) Deadlock, by setting a reasonably large timeout
+        private struct LockToken : IDisposable
         {
-            Clear = 0,
-            Set = 1,
-            Stopped = 2,
+            private object _lockObject;
+
+            public LockToken(object lockObject)
+            {
+                Debug.Assert(lockObject != null);
+
+                _lockObject = lockObject;
+
+                Debug.Assert(!Monitor.IsEntered(_lockObject));
+
+#if DEBUG
+                bool success = Monitor.TryEnter(_lockObject, 10000);
+                Debug.Assert(success);
```

Comment:
Nit: could you add an assert message to at least this one?  If it fires, it'll be helpful to immediately know it was because we timed out waiting for a lock we should never time out waiting for.